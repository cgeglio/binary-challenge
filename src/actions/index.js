export const login = user => ({
  type: 'LOGIN',
  user
})

export const logout = user => ({
  type: 'LOGOUT',
  user
})

export const addCards = cards => ({
  type: 'ADD_CARDS',
  cards
})

export const removeCards = cards => ({
  type: 'REMOVE_CARDS',
  cards
})

export const addQuestion = question => ({
  type: 'ADD_QUESTION',
  question
})

export const removeQuestion = question => ({
  type: 'REMOVE_QUESTION',
  question
})

export const addFortune = fortune => ({
  type: 'ADD_FORTUNE',
  fortune
})

export const removeFortune = fortune => ({
  type: 'REMOVE_FORTUNE',
  fortune
})

export const addReading = currentReading => ({
  type: 'ADD_READING',
  currentReading
})

export const removeReading = currentReading => ({
  type: 'REMOVE_READING',
  currentReading
})

export const addFavorite = favorite => ({
  type: 'ADD_FAVORITE',
  favorite
})

export const removeFavorite = favorite => ({
  type: 'REMOVE_FAVORITE',
  favorite
})

export const resetFavorites = favorites => ({
  type: 'RESET_FAVORITES',
  favorites
})
