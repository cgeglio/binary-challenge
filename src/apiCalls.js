export const getCards = () => {
  let proxyUrl = 'https://cors-anywhere.herokuapp.com/',
      targetUrl = 'https://rws-cards-api.herokuapp.com/api/v1/cards/random?n=4'
  return fetch(proxyUrl + targetUrl)
    .then(response => {
      if (!response.ok) {
        throw Error('Error fetching cards')
      }
      return response.json()
    })
}

export const getFortune = () => {
  return fetch('http://fortunecookieapi.herokuapp.com/v1/fortunes/')
    .then(response => {
      if (!response.ok) {
        throw Error('Error fetching fortune')
      }
      return response.json()
    })
}
