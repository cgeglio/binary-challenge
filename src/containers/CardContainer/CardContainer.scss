@import url('https://fonts.googleapis.com/css?family=Croissant+One&display=swap');
@import "../../mixins.scss";

.card-container {
  @include flex();
}

.question {
  color: #D5A1EE;
  font-family: 'Croissant One';
  margin-bottom: 25px;
}

.cards {
  display: flex;
  flex-direction: row;
  margin: 10px 0px 40px 0px;
}
