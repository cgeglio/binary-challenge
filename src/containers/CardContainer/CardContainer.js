import React, { Component } from 'react';
import './CardContainer.scss';
import { connect } from 'react-redux';
import { getCards, getFortune } from '../../apiCalls';
import { addFavorite, removeFavorite, addCards, addFortune, addReading, removeCards, removeQuestion } from '../../actions';
import Card from '../../components/Card/Card';
import Loader from '../../components/Loader/Loader';
import { Link } from 'react-router-dom';
import save from '../../images/save.png';
import saved from '../../images/saved.png';
import PropTypes from 'prop-types';

export class CardContainer extends Component {
  constructor() {
    super();
    this.state={icon: save}
  }

  componentDidMount() {
    getCards(this.props.spreadNumber)
      .then(cards => this.props.addCardsToStore(cards.cards))
      .then(() => this.fetchFortune())
      .catch(error => console.log(error))
  }

  fetchFortune = () => {
    let fortuneIndex = Math.floor(Math.random() * 100);
    getFortune()
      .then(fortunes => this.props.addFortuneToStore(fortunes[fortuneIndex].message))
      .then(() => this.addCurrentReading())
      .catch(error => console.log(error))
  }

  addCurrentReading = () => {
    let id = Date.now();
    let currentReading = {cards: this.props.cards, fortune: this.props.fortune, question: this.props.question, id: id, saved: false};
    this.props.addReadingToStore(currentReading);
  }

  updateSavedStatus = () => {
    return !this.props.currentReading.saved ? this.saveReading() : this.removeReading();
  }

  saveReading = () => {
    this.setState({icon: saved});
    this.props.currentReading.saved = true;
    this.props.addReadingToFavorites(this.props.currentReading);
  }

  removeReading = () => {
    this.setState({icon: save});
    this.props.currentReading.saved = false;
    this.props.removeReadingFromFavorites(this.props.currentReading);
  }

  determineIcon = () => {
    return this.props.currentReading.saved ? saved : save;
  }

  resetInfo = () => {
    this.props.resetQuestionInStore(this.props.question);
    this.props.removeCards(this.props.cards);
  }

  render() {

    return (
      !this.props.fortune ? <Loader /> :
      <section className='card-container fade-in'>
        <section className='cards'>
          {this.props.cards.map(card => {
            let number = this.props.cards.indexOf(card) + 1;
            return <Card key={card.name_short + card.value} card={card} number={number}/>
          })}
        </section>
        <section className='reading-details'>
          <button onClick={() => this.updateSavedStatus()} className="save-btn"><img src={this.determineIcon()} alt="save reading icon" className="save-icon"/></button>
          <div>
            <h2 className='question'>{this.props.currentReading.question}</h2>
            <h2>{this.props.fortune}</h2>
          </div>
        </section>
        <Link to='/home'><button onClick={() => this.resetInfo()} id='ask-another' className='back-btn'>Ask Another Question</button></Link>
      </section>
    )
  }
}

export const mapStateToProps = state => ({
  cards: state.cards,
  question: state.question,
  fortune: state.fortune,
  currentReading: state.currentReading,
  spreadNumber: state.spreadNumber
})

export const mapDispatchToProps = dispatch => ({
  addCardsToStore: cards => (dispatch(addCards(cards))),
  addFortuneToStore: fortune => (dispatch(addFortune(fortune))),
  addReadingToStore: currentReading => (dispatch(addReading(currentReading))),
  removeCards: cards => (dispatch(removeCards(cards))),
  addReadingToFavorites: favorite => (dispatch(addFavorite(favorite))),
  removeReadingFromFavorites: favorite => (dispatch(removeFavorite(favorite))),
  resetQuestionInStore: question => (dispatch(removeQuestion(question)))
})

export default connect(mapStateToProps, mapDispatchToProps)(CardContainer);

CardContainer.propTypes = {
  cards: PropTypes.array,
  question: PropTypes.string,
  fortune: PropTypes.string,
  currentReading: PropTypes.object,
  addCardsToStore: PropTypes.func,
  addFortuneToStore: PropTypes.func,
  addReadingToStore: PropTypes.func,
  removeCards: PropTypes.func,
  addReadingToFavorites: PropTypes.func,
  removeReadingFromFavorites: PropTypes.func,
  resetQuestionInStore: PropTypes.func,
}
