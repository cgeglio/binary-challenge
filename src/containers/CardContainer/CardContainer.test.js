import React from 'react';
import { CardContainer, mapDispatchToProps, mapStateToProps } from './CardContainer';
import { addFavorite, removeFavorite, addReading, removeQuestion} from '../../actions';
import save from '../../images/save.png';
import saved from '../../images/saved.png';
import { shallow } from 'enzyme';
import { getCards, getFortune } from '../../apiCalls';

jest.mock('../../apiCalls');

describe('CardContainer', () => {

  let wrapper;
  let mockSpreadNumber;
  let mockQuestion;
  let mockCurrentReading;
  let mockAddReading;
  let mockAddFavorite;
  let mockRemoveFavorite;
  let mockRemoveQuestion;

  beforeEach(() => {
    mockSpreadNumber = 2;
    mockQuestion = 'Is the sky blue?'
    mockCurrentReading = {
      cards: [{name: 'The Moon', value: 6}], question: 'Is Bithcuits a good kitty?', fortune: 'Keep your friends close and your enemies closer',
      id: 25,
      saved: false
    };

    mockAddReading = jest.fn().mockImplementation();
    mockAddFavorite = jest.fn().mockImplementation();
    mockRemoveFavorite = jest.fn().mockImplementation();
    mockRemoveQuestion = jest.fn().mockImplementation();

    getCards.mockImplementation(() => {
      return Promise.resolve({cards:[
        {name: 'The Sun', value: 8},
        {name: 'The Hermit', value: 12}
      ]});
    });

    getFortune.mockImplementation(() => {
      return Promise.resolve([{message: 'The early bird gets the worm'}]);
    });

    wrapper = shallow(<CardContainer
       question={mockQuestion}
       currentReading={mockCurrentReading}
       addReadingToStore={mockAddReading}
       addReadingToFavorites={mockAddFavorite}
       removeReadingFromFavorites={mockRemoveFavorite}
       resetQuestionInStore={mockRemoveQuestion}
    />);
  });

  describe('CardContainer container/component', () => {
    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should start with a default state', () => {
      expect(wrapper.state()).toEqual({icon: save, cards: [
        {name: 'The Sun', value: 8},
        {name: 'The Hermit', value: 12}
      ]});
    })

    it('should call getFortune when fetchFortune is called', () => {
      wrapper.instance().fetchFortune();
      expect(getFortune).toHaveBeenCalled();
    })

    it('should call addReadingToStore when addCurrentReading is called', () => {
      global.Date.now = jest.fn().mockImplementation(() => 12345)
      wrapper.instance().setState({cards: [{name: 'The Moon', value: 6}]})
      const expected = {
        cards: [{name: 'The Moon', value: 6}],
        fortune: 'The early bird gets the worm',
        question: mockQuestion,
        id: 12345,
        saved: false
      }
      wrapper.instance().addCurrentReading('The early bird gets the worm');
      expect(mockAddReading).toHaveBeenCalledWith(expected);
    })

    it('should call saveReading if the reading has not been saved yet', () => {
      wrapper.instance().saveReading = jest.fn();
      wrapper.instance().updateSavedStatus();
      expect(wrapper.instance().saveReading).toHaveBeenCalled();
    })

    it('should call removeReading if the reading is already saved', () => {
      mockCurrentReading.saved = true;
      wrapper.instance().removeReading = jest.fn();
      wrapper.instance().updateSavedStatus();
      expect(wrapper.instance().removeReading).toHaveBeenCalled();
    })

    it('should set state, update the current reading\'s saved value, and call addReadingToFavorites when saveReading is invoked', () => {
      const expected = {
        cards: [{name: 'The Moon', value: 6}], question: 'Is Bithcuits a good kitty?', fortune: 'Keep your friends close and your enemies closer',
        id: 25,
        saved: true
      };
      wrapper.instance().saveReading();
      expect(wrapper.state()).toEqual({icon: saved, cards:[
        {name: 'The Sun', value: 8},
        {name: 'The Hermit', value: 12}
      ]})
      expect(mockAddFavorite).toHaveBeenCalledWith(expected);
    })

    it('should set state, update the current reading\'s saved value, and call removeReadingFromFavorites when removeReading is invoked ', () => {
      mockCurrentReading.saved = true;
      const expected = {
        cards: [{name: 'The Moon', value: 6}], question: 'Is Bithcuits a good kitty?', fortune: 'Keep your friends close and your enemies closer',
        id: 25,
        saved: false
      };
      wrapper.instance().removeReading();
      expect(wrapper.state()).toEqual({icon: save, cards:[
        {name: 'The Sun', value: 8},
        {name: 'The Hermit', value: 12}
      ]})
      expect(mockRemoveFavorite).toHaveBeenCalledWith(expected);
    })

    it('should be able to determine which icon to display', () => {
      expect(wrapper.instance().determineIcon()).toEqual(save);
      mockCurrentReading.saved = true;
      expect(wrapper.instance().determineIcon()).toEqual(saved);
    })

    it('should call resetQuestionInStore when resetInfo is called', () => {
      wrapper.instance().resetInfo()
      expect(mockRemoveQuestion).toHaveBeenCalledWith(mockQuestion);
    })

    it('should call updateSavedStatus when the save button is clicked', () => {
      wrapper.instance().updateSavedStatus = jest.fn();
      wrapper.instance().forceUpdate();
      wrapper.find('.save-btn').simulate('click');
      expect(wrapper.instance().updateSavedStatus).toHaveBeenCalled();
    });
  });

  describe('mapStateToProps', () => {
    it('should return a question, a spread number, and a current reading object', () => {
      const mockDispatch = jest.fn();
      const mockState = {
        spreadNumber: 3,
        user: {name: 'Bithcuits', id: 9},
        cards: [{name: 'The Hermit', value: 3}],
        question: 'Is Bithcuits a good kitty?',
        fortune: 'Keep your friends close and your enemies closer',
        currentReading: {
          cards: [{name: 'The Moon', value: 6}],
          question: 'Is Bithcuits a good kitty?',
          fortune: 'Keep your friends close and your enemies closer',
          id: 25,
          saved: false
        }
      }
      const expected = {
        spreadNumber: 3,
        question: 'Is Bithcuits a good kitty?',
        currentReading: {
          cards: [{name: 'The Moon', value: 6}],
          question: 'Is Bithcuits a good kitty?',
          fortune: 'Keep your friends close and your enemies closer',
          id: 25,
          saved: false
        }
      }
      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected);
    })
  });

  describe('mapDispatchToProps', () => {

    it('should call dispatch with the addReading action when addReadingToStore is called', () => {
      const mockReading = {cards: [{name: 'The Magician', value: 9}], id: 8};
      const mockDispatch = jest.fn();
      const actionToDispatch = addReading(mockReading);
      const mappedProps = mapDispatchToProps(mockDispatch);

      mappedProps.addReadingToStore(mockReading);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    })

    it('should call dispatch with the addFavorite action when addReadingToFavorites is called', () => {
      const mockFavorite = {cards: [{name: 'The Magician', value: 9}]};
      const mockDispatch = jest.fn();
      const actionToDispatch = addFavorite(mockFavorite);
      const mappedProps = mapDispatchToProps(mockDispatch);

      mappedProps.addReadingToFavorites(mockFavorite);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    })

    it('should call dispatch with the removeFavorite action when removeReadingFromFavorites is called', () => {
      const mockFavorite = {cards: [{name: 'The Moon', value: 6}]};
      const mockDispatch = jest.fn();
      const actionToDispatch = removeFavorite(mockFavorite);
      const mappedProps = mapDispatchToProps(mockDispatch);

      mappedProps.removeReadingFromFavorites(mockFavorite);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    })

    it('should call dispatch with the removeQuestion action when resetQuestionInStore is called', () => {
      const mockQuestion = 'Is the sky blue?';
      const mockDispatch = jest.fn();
      const actionToDispatch = removeQuestion(mockQuestion);
      const mappedProps = mapDispatchToProps(mockDispatch);

      mappedProps.resetQuestionInStore(mockQuestion);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    })
  });
});
