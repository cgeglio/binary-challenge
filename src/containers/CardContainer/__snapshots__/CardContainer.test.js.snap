// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardContainer CardContainer container/component should match the snapshot 1`] = `
<section
  className="card-container fade-in"
>
  <section
    className="cards"
  >
    <Card
      card={
        Object {
          "name": "The Magician",
          "value": 8,
        }
      }
    />
  </section>
  <section
    className="reading-details"
  >
    <button
      className="save-btn"
      onClick={[Function]}
    >
      <img
        alt="save reading icon"
        className="save-icon"
        src="save.png"
      />
    </button>
    <div>
      <h2
        className="question"
      >
        Is Bithcuits a good kitty?
      </h2>
      <h2>
        Keep your friends close and your enemies closer
      </h2>
    </div>
  </section>
  <Link
    to="/home"
  >
    <button
      className="back-btn"
      id="ask-another"
      onClick={[Function]}
    >
      Ask Another Question
    </button>
  </Link>
</section>
`;
